# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
!unzip data

import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

Archive:  data.zip
   creating: data/
  inflating: data/small_vocab_en     
  inflating: data/small_vocab_fr     


In [0]:
!unzip data

Archive:  data.zip
replace data/small_vocab_en? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: data/small_vocab_en     
replace data/small_vocab_fr? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: data/small_vocab_fr     


## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [0]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[-1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

In [0]:
j

[[4, 5, 9, 6, 7, 8]]

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [0]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    source_id_temp = [[source_vocab_to_int.get(word, 0) for word in sentence.split()]  for sentence in source_text.split('\n')]
    
    target_id_temp = [[target_vocab_to_int.get(word, 0) for word in sentence.split()]  for sentence in target_text.split('\n')]
    
    source_id_text = []
    
    target_id_text = []
    
    #looping through the sentences and adding end of sentence;
    
    for i in range(len(source_id_temp)):
      
      
        source_id_text.append(source_id_temp[i])
      
        target_id_text.append(target_id_temp[i] + [target_vocab_to_int['<EOS>']])
      
     
    
    return (source_id_text, target_id_text)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

In [0]:
source_int_text[:10]

[[18, 41, 83, 33, 89, 230, 91, 206, 53, 50, 83, 67, 207, 69, 17],
 [183, 160, 111, 83, 79, 38, 230, 163, 206, 53, 50, 83, 79, 161, 207, 72, 17],
 [61, 83, 79, 89, 230, 87, 206, 53, 50, 83, 79, 152, 207, 78, 17],
 [183, 160, 111, 83, 33, 56, 230, 78, 206, 53, 50, 83, 23, 207, 82, 17],
 [12, 193, 147, 98, 83, 183, 58, 206, 156, 121, 193, 147, 83, 183, 167, 17],
 [195, 187, 98, 83, 183, 179, 206, 156, 121, 187, 83, 183, 58, 17],
 [171, 83, 116, 230, 105, 206, 156, 50, 83, 79, 38, 207, 163, 17],
 [18, 41, 83, 103, 230, 186, 206, 53, 50, 83, 9, 152, 207, 87, 17],
 [113, 193, 147, 98, 83, 183, 16, 206, 156, 121, 193, 147, 83, 183, 58, 17],
 [183, 160, 111, 83, 33, 103, 230, 221, 206, 53, 50, 83, 33, 42, 207, 72, 17]]



```
# This is formatted as code
```

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [146]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.14.0
Default GPU Device: /device:GPU:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [150]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    
    inputs = tf.placeholder(tf.int32, [None, None], name= "input")
    
    targets= tf.placeholder(tf.int32, [None, None])
    
    learning_rate = tf.placeholder(tf.float32)
    
    keep_prob = tf.placeholder(tf.float32, name = "keep_prob")
    
    target_sequence_length = tf.placeholder(tf.int32, [None], name = "target_sequence_length")
    
    max_target_len  = tf.reduce_max(target_sequence_length, name = "max_target_len")
    
    source_sequence_length = tf.placeholder(tf.int32, [None], name = "source_sequence_length")
    
    
    
    
    return (inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [151]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    
    #remove last word of batch to balance the batch addition at beginning of batch
    
    target_slice = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    
    
    #add GO ID to beginning of batch
    
    target_batch = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int["<GO>"]), target_slice], axis = 1)
      
         
    return target_batch

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [153]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
    
      #build an rnn cell of specific size
        lstm = tf.nn.rnn_cell.BasicLSTMCell(rnn_size)
      
      #wrap with dropout
        drop = tf.nn.rnn_cell.DropoutWrapper(lstm, output_keep_prob = keep_prob)
      
        return drop
    
    
    # Stack up multiple LSTM layers, for deep learning
    cell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    embed = tf.contrib.layers.embed_sequence(rnn_inputs, vocab_size= source_vocab_size, embed_dim= encoding_embedding_size)
    
    
    #creates RNN and performs dynamic unrolling of inputs
    
    RNN_output, RNN_state = tf.nn.dynamic_rnn(cell, embed, dtype = tf.float32)
    
    #This way, dynamic_rnn has an initial hidden state for each sequence. 
    #It will pass on the hidden state from time step to time step for each sequence in the inputs parameter on its own, 
    #and hidden_state_out will contain the final output state for each sequence in the batch to pass to next batch
    #No hidden state is passed between sequences of the same batch, but only between time steps of the same sequence.
    
    
    return (RNN_output, RNN_state)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [154]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target baentch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    
    #decoder cell wrapped with dopout
    
    #dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
    
    #initiliazation of decoder to receive the embedded inputs
    
    decode_helper= tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    
    
    #builds the decoder model. It means it connects the RNN layer(s) on the decoder side and 
    #the input prepared by TrainingHelper.
    
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, decode_helper, initial_state = encoder_state,\
                                                   output_layer = output_layer)
    
    
    #dynamic_decode unrolls the decoder model so that actual prediction can be retrieved by BasicDecoder for each time steps.
    
    dynamic_outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(basic_decoder,\
                                                              maximum_iterations = max_summary_length)
    
    
    
    
    return dynamic_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [155]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    
    #GreedyEmbeddingHelper dynamically takes the output of the current step and give it to the next time step’s input 
    #for inference.In order to embed the each input result dynamically, embedding parameter(just bunch of weight values) 
    #should be provided. Along with it, GreedyEmbeddingHelper asks to give the start_of_sequence_id 
    #for the same amount as the batch size and end_of_sequence_id.
    
    
    decode_helper =  tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,\
                                                              tf.fill([batch_size], start_of_sequence_id), end_of_sequence_id)
    
    
    basic_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, decode_helper, initial_state = encoder_state,\
                                                   output_layer = output_layer)
    
    dynamic_outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(basic_decoder,\
                                                              maximum_iterations = max_target_sequence_length)
    
    
    
    
    return dynamic_outputs


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [156]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    
    #tf.contrib.layers.embed_sequence creates internal representation of embedding parameter, 
    #so we cannot look into or retrieve it.
    #Rather, you need to create a embedding parameter manually by TF Variable.
    
    dec_embedding = tf.Variable(tf.random_uniform((target_vocab_size, decoding_embedding_size), -1, 1))
    
    input_embed = tf.nn.embedding_lookup(dec_embedding, dec_input)
    
    
    # TODO: Implement Function
    def build_cell(rnn_size, keep_prob):
    
      #build an rnn cell of specific size
        lstm = tf.nn.rnn_cell.BasicLSTMCell(rnn_size)
      
      #wrap with dropout
        drop = tf.nn.rnn_cell.DropoutWrapper(lstm, output_keep_prob = keep_prob)
      
        return drop
    
    
    # Stack up multiple LSTM layers, for deep learning
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size, keep_prob) for _ in range(num_layers)])
   
    
    
    with tf.variable_scope("decoder"):
      
      
        output_layer = tf.layers.Dense(units = target_vocab_size)
      
      
        decode_output_train = decoding_layer_train(encoder_state, dec_cell, input_embed, 
                         target_sequence_length, max_target_sequence_length, 
                                                 output_layer, keep_prob)
      
      
      
      
      
    with tf.variable_scope("decoder", reuse = True):
      
      output_layer = tf.layers.Dense(units = target_vocab_size)
      
      decode_output_infer= decoding_layer_infer(encoder_state, dec_cell, dec_embedding, target_vocab_to_int["<GO>"], 
                           target_vocab_to_int["<EOS>"], max_target_sequence_length,
                           target_vocab_size, output_layer, batch_size, keep_prob)
      
    
    
    
  
    
    
    
    
    return decode_output_train, decode_output_infer



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [160]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    #building encoder model
    encoder_output, encoder_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob,\
                              source_sequence_length, source_vocab_size,\
                              enc_embedding_size)
    
    #preprocessing for decoder model
    decoding_target = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    #building training and inference parts of decoder model
    decode_output_train, decode_output_infer = decoding_layer(decoding_target, encoder_state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    
    
    
    
    
    
    return decode_output_train, decode_output_infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [0]:
# Number of Epochs
epochs = 6
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 1.0
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [0]:

import os
#os.mkdir('checkpoints')
save_path = 'checkpoints/dev'

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()


max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()

with train_graph.as_default():
  
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    
    input_shape = tf.shape(input_data)
    
    
    #train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
    
    train_logits, inference_logits = seq2seq_model(input_data,
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    
    
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')
   
  
    #the idea of masks places True on all specified range and false for everything outside the mask
    #specifying the maximum sentence enables the mask form a fixed length of that size.
    
    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

      
   #the idea of masks weights importance on the actual sentence and weights padded sentences with zeros.
   #set all valid timesteps to 1 and all padded timesteps to 0,
  
    with tf.name_scope("optimization"):
      
        # Loss function
     
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        #compute the gradients here
        gradients = optimizer.compute_gradients(cost)
        
        #feed the computed gradients and clip within speciifed value
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        
        #apply the capped gradients updates to optimizer for weight update with capped gradients/put back the gradients 
        #to optimizer
        train_op = optimizer.apply_gradients(capped_gradients)


In [0]:
#import os

#os.mkdir('checkpoints')

os.listdir



<function posix.listdir>

Batch and pad the source and target sequences

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    
    #check the difference between max sentence and sentences, create pads size of the difference
    #add the pads to complete the sentence mength
    
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [187]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    
    #padding incase logits or targets aren't the same size
    #adding pads to the back of target or logit sequence length
    
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1] :
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

            

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   50/1077 - Train Accuracy: 0.3523, Validation Accuracy: 0.4237, Loss: 2.1705
Epoch   0 Batch  100/1077 - Train Accuracy: 0.3867, Validation Accuracy: 0.4439, Loss: 1.3223
Epoch   0 Batch  150/1077 - Train Accuracy: 0.5257, Validation Accuracy: 0.5330, Loss: 0.9481
Epoch   0 Batch  200/1077 - Train Accuracy: 0.5035, Validation Accuracy: 0.5664, Loss: 0.8584
Epoch   0 Batch  250/1077 - Train Accuracy: 0.5692, Validation Accuracy: 0.5792, Loss: 0.6687
Epoch   0 Batch  300/1077 - Train Accuracy: 0.5683, Validation Accuracy: 0.5863, Loss: 0.6794
Epoch   0 Batch  350/1077 - Train Accuracy: 0.5844, Validation Accuracy: 0.5835, Loss: 0.6439
Epoch   0 Batch  400/1077 - Train Accuracy: 0.5941, Validation Accuracy: 0.5966, Loss: 0.5929
Epoch   0 Batch  450/1077 - Train Accuracy: 0.5691, Validation Accuracy: 0.6072, Loss: 0.5364
Epoch   0 Batch  500/1077 - Train Accuracy: 0.5926, Validation Accuracy: 0.6165, Loss: 0.5312
Epoch   0 Batch  550/1077 - Train Accuracy: 0.6188, Validati

In [0]:
x = [1,2 ]

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [0]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [193]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    
    sentence = sentence.lower()
    
    word_ids = [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split()]
    
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


In [192]:
r = "te fkdd J"

r.lower()

'te fkdd j'

## Translate
This will translate `translate_sentence` from English to French.

In [195]:
translate_sentence = 'new jersey is sometimes quiet during autumn , and it is freezing in july .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


Input
  Word Ids:      [162, 229, 25, 91, 35, 58, 173, 6, 226, 177, 25, 146, 118, 117, 161]
  English Words: ['new', 'jersey', 'is', 'sometimes', 'quiet', 'during', 'autumn', ',', 'and', 'it', 'is', 'freezing', 'in', 'july', '.']

Prediction
  Word Ids:      [233, 77, 160, 259, 207, 308, 346, 354, 166, 272, 227, 328, 156, 63, 330, 1]
  French Words: new jersey est parfois calme pendant l' automne , et il gèle en juillet . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.